Added system resources and component dependencies (components/resources) cards

jmgrimes opened the pull request:
- Added `EntityHasResourcesCard` to display resources that are part of a system.
- Added `EntityDependsOnComponentsCard` to display components that are dependencies of a component.
- Added `EntityDependsOnResourcesCard` to display resources that are dependencies of a component.
- Updated the example app to take advantage of these new components.

OrkoHunter reviewed (commented):
Good stuff! Just one comment/question from me.

The two components `DependsOnXCard` where X is `Resources`/`Components` are quite identical actually, and I wonder if it makes sense to combine these two in a generic `DependsOnCard` component. Also, we already have `HasXCard` components where X is `Systems`/`Components`/`Subcomponents`. And this PR adds a new `HasResourcesCard` component. It's probably fine as it is, but maybe keeping `catalog/src/components` directory less inflated makes sense? Totally up for suggestions.

jmgrimes commented:
I do agree.  Initially, I took the tact that maybe several of these components could rely on a central component, since `HasSubcomponents`, `HasSystems`, `HasResources` and others seem to use almost identical logic.  I encountered some trickiness in implementation however, when starting down the depends on entities version, as the header logic again relies on a default component type, and there's no default for the dependsOn relationship.  As such, I followed the existing style.  I might have a look at it today and see if I can rework this into a single central component, or I could open an issue to refactor?

OrkoHunter commented:
Yeah, an issue to refactor works for me! :)

Rugvip commented:
Hi @jmgrimes, since you've looked into this a bit, do you think it would be reasonable to aim for a refactor straight away before merging this? These are a couple of components that I expect a fair number of people to pick up quickly, and we'll be creating unnecessary migrations for people if we release this and then refactor immediately after. Then again a migration of "use X instead of X1, X2, and X3" isn't that bad, so would also be up for releasing this if that's where we're headed.

jmgrimes commented:
Took a shot at a refactor, using a common 'RelatedEntitiesCard'.  Changeset now reflects these four additional changes.

- Added `RelatedEntitesCard` as a base implementation of displaying entities that are related to another entity.
- Refactored `HasComponentsCard` to use base `RelatedEntitiesCard`. Card remains backwards compatible.
- Refactored `HasSubcomponentsCard` to use base `RelatedEntitiesCard`. Card remains backwards compatible.
- Refactored `HasSystemsCard` to use base `RelatedEntitiesCard`. Card remains backwards compatible.

The cards used in other places across the system haven't changed, nor has how they are used changed; only the backing implementation.

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5361-9cd9u55oxlgqtrb9tunlql5l0gb4zrw4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608ae787f9542ab7ae30a0ec

adamdmharvey commented on packages/app/src/components/catalog/EntityPage.tsx:
```diff
@@ -331,6 +345,17 @@ const websiteEntityPage = (
+    <EntityLayout.Route path="/dependencies" title="Dependencies">
+      <Grid container spacing={3} alignItems="stretch">
+        <Grid item md={6}>
+          <EntityDependsOnComponentsCard />
```
I notice this also has a `gridItem` variant. I don't know what's best practice here, but I've seen a ton of cards have this variant and it's explicitly defined in the calls from `EntityPage.tsx`.  Wonder if this should be identified here also? (like it is also on line 490 for `EntityHasResourcesCard`.

jmgrimes replied:
Sure.  Some of the other components, such as the GitHub cards don't have it defined, and it's the default, but I can add it in.  Totally flexible.

adamdmharvey commented on packages/app/src/components/catalog/EntityPage.tsx:
```diff
@@ -331,6 +345,17 @@ const websiteEntityPage = (
+          <EntityDependsOnComponentsCard />
+        </Grid>
+        <Grid item md={6}>
+          <EntityDependsOnResourcesCard />
```
Repeat question on `gridItem` variant?

adamdmharvey commented on packages/app/src/components/catalog/EntityPage.tsx:
```diff
@@ -287,6 +290,17 @@ const serviceEntityPage = (
+    <EntityLayout.Route path="/dependencies" title="Dependencies">
+      <Grid container spacing={3} alignItems="stretch">
+        <Grid item md={6}>
+          <EntityDependsOnComponentsCard />
+        </Grid>
+        <Grid item md={6}>
+          <EntityDependsOnResourcesCard />
```
Repeat question regarding `gridItem` variants?

Again, seems pretty duplicative to have to put it everywhere, but it seems to be a common pattern in here?

adamdmharvey reviewed (commented):
Great refactor.  A minor query on the `gridItem` variants?

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5361-9cd9u55oxlgqtrb9tunlql5l0gb4zrw4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608ae787f9542ab7ae30a0ec

jmgrimes commented:
Did a little more monkeying with this to add type parameterization and type safety between the `columns` and the `asRenderableEntities` properties, inspired from the `EntityTable` implementation.  Also renamed constants to presets, in line with the way `EntityTable` handles this.

I noticed that the api-docs plugin does some of this as well, but it feels like crossing from that plugin depending on the catalog plugin feels wrong, yet this lower level shared component seems to fit somewhere in the middle of both.  I left it alone for those reasons, but it might be worth a discussion about where this should go to make its features available to the `ProvidingComponentsCard` and `ConsumingComponentsCard` in that plugin as well.

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr5361-9cd9u55oxlgqtrb9tunlql5l0gb4zrw4.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/608ae787f9542ab7ae30a0ec

jmgrimes commented:
@Rugvip Any further feedback on this pull request after refactor?

freben commented:
Hey, I'm the support goalie currently - it's on my queue here! :)

freben commented on plugins/catalog/src/components/RelatedEntitiesCard/RelatedEntitiesCard.tsx:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  return (
+    <EntityTable
```
Whoa I just now remembered / realized that the table actually just sort of mimics the look of a card. And the paddings and stuff are different.

If it didn't, it would have been neater to render the InfoCard in the outer component (e.g. `HasComponentsCard`), and make this inner component `RelatedEntitiesTable` with fewer props.

OK, something to look into another day. Doesn't affect this review :)

freben commented on plugins/catalog/src/components/RelatedEntitiesCard/RelatedEntitiesCard.tsx:
```diff
@@ -0,0 +1,97 @@
+      <InfoCard variant={variant} title={title}>
+        <WarningPanel
+          severity="error"
+          title={`Could not load ${{ title }}`}
```
accidental double curly braces?

jmgrimes replied:
Yeah, let me go fix that...

freben commented on plugins/catalog/src/components/HasResourcesCard/HasResourcesCard.tsx:
```diff
@@ -0,0 +1,43 @@
+      relationType={RELATION_HAS_PART}
+      columns={resourceEntityColumns}
+      asRenderableEntities={asResourceEntities}
+      emptyMessage="No resource is part of this system"
```
just noting that these cards make assumptions about what the _source_ of the relation is too. That may reduce their utility.

Would it make sense to grab the entity the relations start from, and make the text something like ``` This ${entity.kind.toLocaleLowercase('en-US')} has no associated resources ``` or something like that?

jmgrimes replied:
For sure.  On the one hand, the initial versions of these cards did too, preferring the semantic to be "sort of" encoded in the relationship.  I tried to follow the same semantic in naming, and as it turned out, HasResources only applies to a system.  HasComponents *could* apply to a system or to a component (via subcomponents), but there were both HasComponents and HasSubcomponents defined.  It gets even more confusing when adding cards for DependsOnComponent.  All of these cards do the same thing, and only this message or title varies.

jmgrimes replied:
I'm pretty much open to suggestions here, but changing from the `RelationshipnameTargetkindCard` to the `SourcekindRelationpurposeMaybetargetkindCard` semantic will introduce breaking changes for apps relying on the existing `HasSystemsCard`, `HasComponentsCard`, and `HasSubcomponentsCard`.

What feels like the right approach to you?

freben commented on plugins/catalog/src/components/HasResourcesCard/HasResourcesCard.test.tsx:
```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import { HasResourcesCard } from './HasResourcesCard';
+
+describe('<HasComponentsCard />', () => {
```
```suggestion
describe('<HasResourcesCard />', () => {
```

freben commented on plugins/catalog/src/components/HasComponentsCard/HasComponentsCard.tsx:
```diff
@@ -14,79 +14,30 @@
+      entityKind="Component"
+      relationType={RELATION_HAS_PART}
+      columns={componentEntityColumns}
+      emptyMessage="No component is part of this system"
```
yeah this is another place where the relation can mean different things depending on the context - `Component hasPart Component` means that the first is an uber-component (monorepo style massive apps etc) and the latter is a smaller part of it.

This text shows the complexity of it... https://github.com/backstage/backstage/blob/master/docs/features/software-catalog/well-known-relations.md#partof-and-haspart

freben reviewed (commented):
Hey and sorry for the delay. Lots of great contributions to look through! 😅 

On the one hand this is basically mergeable. But the question about semantic correctness is what I wanted to discuss with you. Essentially, is it even _possible_ to make a `RelationshipnameTargetkindCard` that doesn't have knowledge of the source kind too?

What I'm thinking over here is, is it necessary to make semantic cards `SourcekindRelationpurposeMaybetargetkindCard` instead?

freben commented:
You know, way back when we made the relations initially, my gut feeling for how to visualize them in the frontend was not as individual cards, but as a single list grouped by relation and target kind.

## Part of
- [system icon] system-name-1
## Exposes API
- [api icon] api-name-1
## Has subcomponents
- [component icon] component-name-1
- [component icon] component-name-2
## Depends on
- [component icon] component-name-3

But more on an icon list form using features from https://material-ui.com/components/lists/ and maybe something minimal in terms of what data to show per line item. Could even be a hybrid data table form, absolutely, but all collected in one place - and with ONE such card per SOURCE kind, such that it can contain custom logic for what the texts should be for each heading because it will know both the source and the target at that point and can make sense of them.

I'm open to ideas, but I'm overall wondering if a large number of individual cards with tables really scales as well.

jmgrimes commented:
Sure.  On the one hand, large numbers of individual cards is a pain, because there's a constant influx of needing to build new ones for each From -> Relationship -> To ultimately.  A higher order card would be nice to encapsulate a lot of this.  On the other, a larger encapsulating card doesn't support as much variance in layouts from the application side.  The smaller components do.

I can take this in either direction.  Would you like me to try and take a step at evolving these to more generally support source entity cards?

freben commented:
I'm tempted to have this merged and then look at something more collected separately. Both have their merits and ultimately the end user chooses which they prefer to use

jmgrimes commented:
Happy to take this in either direction.  Let me see if there's any way to make the source entity and relationship be passed in to reduce the number of properties.

jmgrimes commented:
I think I'm going to leave it as is for now, with respect to the passing in of parameters to the internal component.  With the refactor done, it can be evolved without impacting the "public" components already provided, as well as the new components this pull request adds.

freben commented on plugins/catalog/src/components/RelatedEntitiesCard/RelatedEntitiesCard.tsx:
```diff
@@ -0,0 +1,97 @@
+  if (error || !entities) {
+    return (
+      <InfoCard variant={variant} title={title}>
+        <WarningPanel
```
One final ask! Could you make this one use the `ResponseErrorPanel` from `@backstage/core` to render the error instead of a plain `WarningPanel`? It is specialised for this kind of thing, such that is shows rich error info including stack traces carried over from the backend and stuff. It doesn't have a customisable title, but if you wanted that text present you could show it separately above it, or whatever.

freben reviewed (commented):
Yep I think this is great! One fffffiiiinal thing that I missed the first time around, then let's 🚢

jmgrimes commented on plugins/catalog/src/components/RelatedEntitiesCard/presets.ts:
```diff
@@ -0,0 +1,60 @@
+  'https://backstage.io/docs/features/software-catalog/descriptor-format#kind-component';
+export const asComponentEntities = (
+  entities: Entity[] | undefined,
+): ComponentEntity[] => entities as ComponentEntity[];
```
Yeah, it needs to accept an undefined and produce an empty array.  Theoretically, it shouldn't be undefined if error isn't set, but the statement if(error || !entities) covered that previously.  Using the ResponseErrorPanel, I had to remove the undefined check for entities.  This led to breakage in the table with the asRenderableEntities method.  I'll rework that to return an empty array, and get the empty table response.

freben reviewed (approved):
Let's go!
